src/engine/strat_engine/setup.rs

```diff
@@ -26,6 +26,67 @@ use super::range_alloc::RangeAllocator;
 use super::serde_structs::PoolSave;
 
 
+/// Determine if devnode is a stratis device, if it is we will add  pool uuid and device
+/// information to pool_map and return pool uuid.
+pub fn is_stratis_device(pool_map: &mut HashMap<PoolUuid, HashMap<Device, PathBuf>>,
+                         devnode: PathBuf,
+                         devno: u64)
+                         -> EngineResult<Option<PoolUuid>> {
+
+    let f = OpenOptions::new().read(true).open(&devnode);
+
+    // There are some reasons for OpenOptions::open() to return an error
+    // which are not reasons for this method to return an error.
+    // Try to distinguish. Non-error conditions are:
+    //
+    // 1. ENXIO: The device does not exist anymore. This means that the device
+    // was volatile for some reason; in that case it can not belong to
+    // Stratis so it is safe to ignore it.
+    //
+    // 2. ENOMEDIUM: The device has no medium. An example of this case is an
+    // empty optical drive.
+    //
+    // Note that it is better to be conservative and return with an
+    // error in any case where failure to read the device could result
+    // in bad data for Stratis. Additional exceptions may be added,
+    // but only with a complete justification.
+    if f.is_err() {
+        let err = f.unwrap_err();
+        match err.kind() {
+            ErrorKind::NotFound => {
+                return Ok(None);
+            }
+            _ => {
+                if let Some(errno) = err.raw_os_error() {
+                    match Errno::from_i32(errno) {
+                        Errno::ENXIO | Errno::ENOMEDIUM => return Ok(None),
+                        _ => return Err(EngineError::Io(err)),
+                    };
+                } else {
+                    return Err(EngineError::Io(err));
+                }
+            }
+        }
+    }
+
+    let mut f = f.expect("unreachable if f is err");
+    if let DevOwnership::Ours(pool_uuid, _) = StaticHeader::determine_ownership(&mut f)? {
+        // No value should ever be ejected, because duplicate device nodes
+        // are filtered out above. Therefore, the return value of insert()
+        // might as well be ignored.
+        debug!("STRATIS DEVICE {:?} {:?}", pool_uuid, devnode);
+
+        let _ = pool_map
+            .entry(pool_uuid)
+            .or_insert_with(HashMap::new)
+            .insert(Device::from(devno), devnode);
```

Comment:
It was probably cowardly to have a ```let _``` instead of an ```expect()``` here, see comment above. But now, the circumstance on which the comment relies are no longer true, so the code needs some modfication. Given these changes, I would _not_ pass the HashMap, instead I'ld just return the relevant values.

Reply:
If I return the relevant values I will need to duplicate the same code to insert them into the hash map, is that really better?  Yes the comment is wrong now.

Reply:
I'm not sure what you mean by duplicate. Won't the insertion code be in just one place?

Reply:
There are two different code paths which call `is_stratis_device`.  If each of them needs the return values placed into a hashmap there will be duplicate insertion code for each to do so.